src/libs/safe-path/src/safe_join.rs

```diff
@@ -0,0 +1,262 @@
+///
+/// # Arguments
+/// - `root`: the absolute path to scope the symlink evaluation.
+/// - `unsafe_path`: the path to evaluated and joint with `root`. It is unsafe since it may try to
+///   escape from the `root` by using "../" or symlinks.
+///
+/// # Security
+/// On success return, the `safe_join()` function guarantees that:
+/// - The resulting string must be a child path of `root` and will not contain any symlink path
+///   components (they will all be expanded).
+/// - When expanding symlinks, all symlink path components must be resolved relative to the provided
+///   `root`. In particular, this can be considered a userspace implementation of how chroot(2)
+///    operates on file paths.
+/// - Non-existent path components are unaffected.
+///
+/// Note that the guarantees provided by this function only apply if the path components in the
+/// returned string are not modified (in other words are not replaced with symlinks on the
+/// filesystem) after this function has returned. You may use [crate::SafePathBuf] to protect from
+/// such TOCTOU attacks.
+pub fn safe_join<R: AsRef<Path>, U: AsRef<Path>>(root: R, unsafe_path: U) -> Result<PathBuf> {
+    // Assume that `rootfs` exists.
+    let root = root.as_ref().canonicalize()?;
+    if !root.is_absolute() {
+        return Err(Error::new(
+            ErrorKind::Other,
+            format!("Invalid rootfs path: {}", root.display()),
+        ));
+    }
+
+    let mut nlinks = 0u32;
+    let mut curr_path = unsafe_path.as_ref().to_path_buf();
+    'restart: loop {
+        let mut subpath = PathBuf::new();
+        let mut iter = curr_path.components();
+        'next_comp: while let Some(comp) = iter.next() {
+            match comp {
+                Component::Prefix(_) => {
+                    return Err(Error::new(
+                        ErrorKind::Other,
+                        format!("Invalid path prefix in: {}", unsafe_path.as_ref().display()),
+                    ));
+                }
+                Component::RootDir | Component::CurDir => {
+                    continue 'next_comp;
+                }
+                Component::ParentDir => {
+                    subpath.pop();
+                }
+                Component::Normal(n) => {
+                    subpath.push(n);
+                    let path = root.join(&subpath);
+                    if let Ok(v) = path.read_link() {
+                        nlinks += 1;
+                        if nlinks > 255 {
+                            return Err(Error::new(
+                                ErrorKind::Other,
+                                format!(
+                                    "Too many levels of symlinks: {}",
+                                    unsafe_path.as_ref().display()
+                                ),
+                            ));
+                        }
+                        curr_path = if v.is_absolute() {
+                            v.join(iter.as_path())
```

Comment:
What does joining `iter.as_path()` mean? I tested with running some sample code that it will only append a `/` at the end of the path, `iter.as_path()` will always be empty strnig.